mailpoet/lib/Automation/Engine/Storage/WorkflowStorage.php

```diff
@@ -10,43 +10,80 @@
 
 class WorkflowStorage {
   /** @var string */
-  private $table;
+  private $workflowTable;
+  /** @var string */
+  private $versionsTable;
 
   /** @var wpdb */
   private $wpdb;
 
   public function __construct() {
     global $wpdb;
-    $this->table = $wpdb->prefix . 'mailpoet_workflows';
+    $this->workflowTable = $wpdb->prefix . 'mailpoet_workflows';
+    $this->versionsTable = $wpdb->prefix . 'mailpoet_workflow_versions';
     $this->wpdb = $wpdb;
   }
 
   public function createWorkflow(Workflow $workflow): int {
-    $result = $this->wpdb->insert($this->table, $workflow->toArray());
+    $result = $this->wpdb->insert($this->workflowTable, $this->workflowHeaderData($workflow));
     if (!$result) {
       throw Exceptions::databaseError($this->wpdb->last_error);
     }
-    return $this->wpdb->insert_id;
+    $id = (int)$this->wpdb->insert_id;
+    $this->insertWorkflowVersion($id, $workflow);
+    return $id;
   }
 
   public function updateWorkflow(Workflow $workflow): void {
-    $result = $this->wpdb->update($this->table, $workflow->toArray(), ['id' => $workflow->getId()]);
+    $result = $this->wpdb->update($this->workflowTable, $this->workflowHeaderData($workflow), ['id' => $workflow->getId()]);
     if ($result === false) {
       throw Exceptions::databaseError($this->wpdb->last_error);
     }
+    $this->insertWorkflowVersion($workflow->getId(), $workflow);
   }
 
   public function getWorkflow(int $id): ?Workflow {
-    $table = esc_sql($this->table);
-    $query = (string)$this->wpdb->prepare("SELECT * FROM $table WHERE id = %d", $id);
+    $workflowTable = esc_sql($this->workflowTable);
+    $versionTable = esc_sql($this->versionsTable);
+    $query = (string)$this->wpdb->prepare("
+SELECT
+      workflow.*,
+      version.id AS version_id,
+      version.steps,
+      version.trigger_keys
+FROM
+      $workflowTable as workflow,
+      $versionTable as version
+WHERE
+      version.workflow_id = workflow.id
+ORDER BY
+      version.id DESC
+LIMIT
+      0,1; ",
+      $id
+    );
     $data = $this->wpdb->get_row($query, ARRAY_A);
     return $data ? Workflow::fromArray((array)$data) : null;
   }
 
   /** @return Workflow[] */
   public function getWorkflows(): array {
-    $table = esc_sql($this->table);
-    $query = "SELECT * FROM $table ORDER BY id DESC";
+    $workflowTable = esc_sql($this->workflowTable);
+    $versionTable = esc_sql($this->versionsTable);
+    $query = "
+SELECT
+      workflow.*,
+      version.id AS version_id,
+      version.steps,
+      version.trigger_keys
+FROM
+      $workflowTable AS workflow
+        LEFT JOIN
+         $versionTable as version ON (version.workflow_id=workflow.id)
+WHERE
+      version.id = (SELECT Max(id) FROM $versionTable WHERE workflow_id= version.workflow_id)
+ORDER BY
+      workflow.id DESC; ";
```

Comment:
I was checking where we're using this, and it seems to be only [the `WorkflowsGetEndpoint`](https://github.com/mailpoet/mailpoet/blob/c8278dde2a066f1f4602ef32151cc8d4ad22ca90/mailpoet/lib/Automation/Engine/Endpoints/Workflows/WorkflowsGetEndpoint.php). Considering it's a "listing" endpoint and [seeing the data it uses now](https://github.com/mailpoet/mailpoet/blob/c8278dde2a066f1f4602ef32151cc8d4ad22ca90/mailpoet/lib/Automation/Engine/Endpoints/Workflows/WorkflowsGetEndpoint.php#L29), I suppose we could just keep this method touch only the `workflows` table, and maybe rename it to `getWorkflowsList` to indicate it doesn't return full workflows, only the listing data.

Reply:
I see why it would be beneficial to do so. On the other side, we switch explicitly from `LEFT JOIN` to `INNER JOIN` in the other queries to prevent invalid records being returned. And I think we should follow this thought here too. What do you think?

Reply:
@websupporter Ahh, right, we still need to use the `Workflow` data type. OK, let's keep it consistent. We can optimize it later by adding a read-only dedicated data type, such as `WorkflowListRecord`.